sql/core/src/main/scala/org/apache/spark/sql/execution/SparkPlan.scala

```diff
@@ -220,7 +222,61 @@ abstract class SparkPlan extends QueryPlan[SparkPlan] with Logging with Serializ
    * Runs this query returning the result as an array.
    */
   def executeCollect(): Array[InternalRow] = {
-    execute().map(_.copy()).collect()
+    // Packing the UnsafeRows into byte array for faster serialization.
+    // The byte arrays are in the following format:
+    // [size] [bytes of UnsafeRow] [size] [bytes of UnsafeRow] ... [-1]
+    val byteArrayRdd = execute().mapPartitionsInternal { iter =>
+      new Iterator[Array[Byte]] {
```

Comment:
i also find this more understandable if you just write it imperatively within the map partitions; something like

```scala
execute().mapPartitionsInternal { iter =>
  while (iter.hasNext) {
    // write each row to a buffer
  }
  Iterator(buffer)
}
```